### ISSUE TITLE
Implement min_precision to merge numbers

### ISSUE THREAD

**Issue by Laouen** [BoostGSoC18/Real#11]
As done in path compression, whenever a number iterator is increased and the calculated number entire representation is reached (lower bound is equal to upper bound), then the real number is converted to a real_explicit with the calculated number and the children are destroyed. 

This can improve performance to reduce both memory and recursive calls

**Comment by sdavtaker** [comment]
For this to work the leaves below that node have to all be real_explicit

**Comment by Laouen** [comment]
That's correct, but once real_operation is merged and it has been converted to a single real_explicit, then the parent real_operation node could also be merged, in the best case scenario, the whole tree is reduced to a single real_explicit.

Similar to path compression, this can be made when an iterator is increased and its two boundaries (lower and upper) have the exact same value, if that is the case, the iterator has reached the number full precision and the number can be merged. For this purpose, we can declare a non-const iterator, an iterator that has this behaviour as a side effect. The current iterator is const and cannot modify the pointed number. The user will choose which one fits best for his needs.
